tools/fast_build_tar/buildtar.go

```diff
@@ -0,0 +1,540 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"archive/tar"
+	"bufio"
+	"compress/bzip2"
+	"compress/gzip"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/golang/glog"
+
+	"golang.org/x/build/pargzip"
+)
+
+type tarFile struct {
+	directory string
+
+	tw *tar.Writer
+
+	meta     fileMeta
+	dirsMade map[string]struct{}
+
+	closers []func()
+}
+
+func newTarFile(output, directory, compression string, meta fileMeta) (*tarFile, error) {
+	var (
+		w       io.Writer
+		closers []func()
+	)
+	f, err := os.Create(output)
+	if err != nil {
+		return nil, err
+	}
+	closers = append(closers, func() {
+		f.Close()
+	})
+	w = f
+
+	buf := bufio.NewWriter(w)
+	closers = append(closers, func() { buf.Flush() })
+	w = buf
+
+	switch compression {
+	case "":
+	case "gz":
+		gzw := pargzip.NewWriter(w)
+		closers = append(closers, func() { gzw.Close() })
+		w = gzw
+	case "bz2", "xz":
+		return nil, fmt.Errorf("%q compression is not supported yet. File a bug to get it implemented!", compression)
+	default:
+		return nil, fmt.Errorf("unknown compression %q", compression)
+	}
+
+	return &tarFile{
+		directory: directory,
+		tw:        tar.NewWriter(w),
+		closers:   closers,
+		meta:      meta,
+		dirsMade:  map[string]struct{}{},
+	}, nil
+}
+
+func (f *tarFile) addFile(file, dest string) error {
+	dest = strings.TrimLeft(dest, "/")
+	dest = filepath.Clean(dest)
+
+	uid := f.meta.getUid(dest)
+	gid := f.meta.getGid(dest)
+	uname := f.meta.getUname(dest)
+	gname := f.meta.getGname(dest)
+
+	dest = filepath.Join(strings.TrimLeft(f.directory, "/"), dest)
+	dest = filepath.Clean(dest)
+
+	info, err := os.Stat(file)
+	if err != nil {
+		return err
+	}
+
+	mode := f.meta.getMode(dest)
+	// If mode is unspecified, derive the mode from the file's mode.
+	if mode == 0 {
+		mode = os.FileMode(0644)
+		if info.Mode().Perm()&os.FileMode(0111) != 0 {
+			mode = os.FileMode(0755)
+		}
+	}
+
+	header := tar.Header{
+		Name:       dest,
+		Mode:       int64(mode),
+		Uid:        uid,
+		Gid:        gid,
+		Size:       0,
+		Uname:      uname,
+		Gname:      gname,
+		AccessTime: time.Now(),
```

Comment:
I'm pretty sure the reference version always uses a timestamp of 0 (for reproducibility).